### ISSUE TITLE
Document speclib format

### ISSUE THREAD

**Issue by chambm** [vdemichev/DiaNN#253]
Hi @vdemichev , could you release an updated schema for the speclib format for the current speclib version? It doesn't have to be fancy, just a list of fields, their purpose, and their data types/sizes. Or open source the speclib reading/writing part of the code? Thanks.

**Comment by vdemichev** [comment]
Hi Matt,

Thanks for the reminder, below is the full specification. Hope this helps, and many thanks for implementing the converter! Please let me know if something is missing.

Best,
Vadim

`
// first templates for read/write functions

template <class F, class T> void write_vector(F &out, std::vector<T> &v) {
	int size = v.size();
	out.write((char*)&size, sizeof(int));
	if (size) out.write((char*)&(v[0]), size * sizeof(T));
}

template <class F, class T> void read_vector(F &in, std::vector<T> &v) {
	int size = 0; in.read((char*)&size, sizeof(int));
	if (size) {
		v.resize(size);
		in.read((char*)&(v[0]), size * sizeof(T));
	}
}

template<class F> void write_string(F &out, std::string &s) {
	int size = s.size();
	out.write((char*)&size, sizeof(int));
	if (size) out.write((char*)&(s[0]), size);
}

template<class F> void read_string(F &in, std::string &s) {
	int size = 0; in.read((char*)&size, sizeof(int));
	if (size) {
		s.resize(size);
		in.read((char*)&(s[0]), size);
	}
}

template <class F, class T> void write_array(F &out, std::vector<T> &a) {
	int size = a.size();
	out.write((char*)&size, sizeof(int));
	for (int i = 0; i < size; i++) a[i].write(out);
}

template <class F, class T> void write_array(F& out, std::vector<T>& a, int version) {
	int size = a.size();
	out.write((char*)& size, sizeof(int));
	for (int i = 0; i < size; i++) a[i].write(out, version);
}

template <class F, class T> void read_array(F &in, std::vector<T> &a) {
	int size = 0; in.read((char*)&size, sizeof(int)); 
	if (size) {
		a.resize(size);
		for (int i = 0; i < size; i++) a[i].read(in);
	}
}

template <class F, class T> void read_array(F& in, std::vector<T>& a, int version) {
	int size = 0; in.read((char*)& size, sizeof(int));
	if (size) {
		a.resize(size);
		for (int i = 0; i < size; i++) a[i].read(in, version);
	}
}

template<class F> void write_strings(F &out, std::vector<std::string> &strs) {
	int size = strs.size();
	out.write((char*)&size, sizeof(int));
	for (int i = 0; i < size; i++) write_string(out, strs[i]);
}

template<class F> void read_strings(F &in, std::vector<std::string> &strs) {
	int size = 0; in.read((char*)&size, sizeof(int));
	if (size) {
		strs.resize(size);
		for (int i = 0; i < size; i++) read_string(in, strs[i]);
	}
}

// fragment ion struct
class Product {
public:
	float mz = 0.0;
	float height = 0.0;
	uint8_t charge = 0, type = 0, index = 0, loss = 0;
}

// read a charged peptide (either a target or a decoy)
template <class F> void Peptide::read(F &in, int v) {
	in.read((char*)&index, sizeof(int));
	in.read((char*)&charge, sizeof(int));
	in.read((char*)&length, sizeof(int));

	in.read((char*)&mz, sizeof(float));
	in.read((char*)&iRT, sizeof(float));
	in.read((char*)&sRT, sizeof(float));

	if (v <= -2) {
		in.read((char*)& lib_qvalue, sizeof(float));
		in.read((char*)& iIM, sizeof(float));
		in.read((char*)& sIM, sizeof(float));
	}

	read_vector(in, fragments);
}

// read a library entry
template <class F> void Entry::read(F &in, int v) {
	target.read(in, v); // target is a Peptide here, see the function above
	int dc = 0; in.read((char*)&dc, sizeof(int));
	if (dc) decoy.read(in, v); // decoy is a Peptide here, see the function above

	int ff = 0, prt = 0;
	in.read((char*)&ff, sizeof(int));
	in.read((char*)&prt, sizeof(int));
	entry_flags = ff, proteotypic = prt;
	in.read((char*)&pid_index, sizeof(int));
	read_string(in, name);
	if (v <= -3) {
		in.read((char*)&pg_qvalue, sizeof(float));
		in.read((char*)&ptm_qvalue, sizeof(float));
		in.read((char*)&site_conf, sizeof(float));
	}
}

// read .speclib
template<class F> void read(F &in) {
	int gd = 0, gc = 0, ip = 0, version = 0;

	in.read((char*)&version, sizeof(int));
	if (version >= 0) gd = version;
	else in.read((char*)&gd, sizeof(int));
	if (version < -3) {
		dsout << "ERROR: version " << -version << " of the .speclib format is not supported by this DIA-NN version\n";
		exit(-1);
	}

	in.read((char*)&gc, sizeof(int));
	in.read((char*)&ip, sizeof(int));
	gen_decoys = gd, gen_charges = gc, infer_proteotypicity = ip;

	read_string(in, name);
	read_string(in, fasta_names);
	read_array(in, proteins);
	read_array(in, protein_ids);
	read_strings(in, precursors);
	read_strings(in, names);
	read_strings(in, genes);
	in.read((char*)&iRT_min, sizeof(double));
	in.read((char*)&iRT_max, sizeof(double));
	read_array(in, entries, version);
	for (auto &e : entries) e.lib = this;
	if (version <= -1 && in.peek() != std::char_traits<char>::eof()) read_vector(in, elution_groups); // elution_groups is a vector of type int
}
`
